src/installation/guides/efi-stub.md

```diff
@@ -0,0 +1,162 @@
+# EFI Boot Stub
+
+When the kernel is compiled with the [EFI boot
+stub](https://www.kernel.org/doc/Documentation/efi-stub.txt) like the one
+packaged in Void, UEFI firmware can load the kernel directly as an EFI
+executable, eliminating the need for a bootloader.
+
+In the following, we set up a simple Void system on a drive, with an encrypted
+root partition and an unencrypted EFI system partition. We will refer to the
+drive as `/dev/sd<X>`.
+
+## Provisioning
+
+On the drive, create a GPT partition table with the following partitions:
+
+| No. | Size              | Type             |
+|-----|-------------------|------------------|
+| 1   | 512MiB            | EFI System       |
+| 2   | rest of the drive | Linux filesystem |
+
+Create a FAT 32 filesystem on the EFI System Partition, optionally giving it a
+label.
+
+```
+# mkfs.fat -F 32 -n efi /dev/sd<X>1
+```
+
+Initialize the LUKS device.
+
+```
+# cryptsetup luksFormat /dev/sd<X>2 --type luks2 --label luks
+```
+
+Open the LUKS device.
+
+```
+# cryptsetup open --type luks /dev/sd<X>2 root
+```
+
+Create the desired file system.
+
+```
+# mkfs.ext4 -L root /dev/mapper/root
+```
+
+## Installation
+
+With the LUKS device still open, mount the root filesystem, the EFI system
+partition and the necessary directories to `/mnt`.
+
+```
+# mount /dev/mapper/root /mnt
+# mkdir /mnt/boot
+# mount /dev/sd<X>1 /mnt/boot
+# for dir in dev proc sys run; do
+> mkdir /mnt/$dir
+> mount --rbind --make-rslave /$dir /mnt/$dir
+> done
+```
+
+Install the Void Linux base system, `cryptsetup` and `efibootmgr`.
+
+```
+# xbps-install -S -R https://alpha.de.repo.voidlinux.org/current -r /mnt base-system cryptsetup efibootmgr
+```
+
+Change the filesystem root to enter the new system.
+
+```
+# chroot /mnt
+```
+
+Set the filesystem permissions, root password and hostname.
+
+```
+# chown root:root /
+# chmod 755 /
+# passwd root
+# echo void > /etc/hostname
+```
+
+Void Linux uses [dracut](https://lwn.net/Articles/317793/) to manage early
+userspace. Configure it to open and mount the LUKS device and mount the EFI
+variables and system partition. To this end, edit `/etc/crypttab`.
+
+```
+# <name>    <device>                  <password>  <options>
+root        /dev/disk/by-uuid/<UUID>  none
+```
```

Comment:
I believe that when `cryptsetup` is installed, dracut pulls in the `crypt` module which automatically looks for LUKS filesystems. It will refer to `/etc/crypttab` for the mapper name, password and options.

Reply:
Those are dracut's LUKS-spefic cmdline arguments: https://mirrors.edge.kernel.org/pub/linux/utils/boot/dracut/dracut.html#_crypto_luks

Reply:
Yeah, I've used the dracut cmdline arguments, and didn't specify anything in crypttab. But if dracut can figure them out from crypttab, that should make life easier, I think. You have to include crypttab in the initramfs in this case, right?

Reply:
I have a bit different setup.
I used EFISTUB in the past and rEFInd now.
With Secure Boot on, and using my own keys.

I think this setup is a bit too complicated and hackish.
This is my configuration:
```
$ grep -v '^#' /etc/fstab
tmpfs			/tmp	tmpfs	defaults,nosuid,nodev			0	0
/dev/vg1/root		/	ext4	defaults,relatime,errors=remount-ro	0	1
/dev/vg1/home		/home	ext4	rw,relatime				0	2
UUID="101C-44EA"	/boot	vfat	rw,noauto,noatime,user,async,dev	0	2
/dev/vg1/swap		swap	swap	defaults,noatime			0	0
/home/danh/Music	/var/lib/mpd/music none bind				0	0
$ grep -v '^#' /etc/crypttab # empty
$ cat /etc/dracut.conf.d/*
hostonly="yes"
tmpdir=/tmp
early_microcode="yes"
omit_dracutmodules+=" bash "
add_dracutmodules+=" dash "
kernel_cmdline+=" rd.luks.key=/etc/keys/luks-key.gpg:/ "
install_items+=" /etc/keys/luks-key.gpg "
kernel_cmdline+=" page_poison=1 "
kernel_cmdline+=" net.ifnames=0 "
kernel_cmdline+=" i915.modeset=1 "
add_dracutmodules+=" crypt-gpg "
```

Reply:
@sgn how are you signing your kernel?

Reply:
@ericonr Create my own keys and add it into firmware, delete M$ keys, Sakaki has a good guideline for it https://wiki.gentoo.org/wiki/Sakaki%27s_EFI_Install_Guide/Configuring_Secure_Boot_under_OpenRC

Sign automatically with https://github.com/void-linux/void-packages/blob/master/srcpkgs/sbsigntool/files/kernel.d/sbsigntool.post-install

Reply:
dracut has a feature to sign UEFI bundles, if you create them with `dracut --uefi`. Just need to put the key location in a dracut config file.

RIght now I'm using https://github.com/Foxboron/sbctl , which I want to package as soon as we get a release.

Reply:
Didn't realize we had a hook too, thanks!

Reply:
> dracut has a feature to sign UEFI bundles, if you create them with `dracut --uefi`. Just need to put the key location in a dracut config file.

Do you mean `dracut`'s `modsign`, which uses `keyctl`?
IIRC, you opened a PR for it, right?
If it's not merged, I could take a look next week.

I'm not really a fan of that approach, though.
Creating the bundle doesn't work well with other hooks that is in place, e.g. rEFInd, efimgr. No?
Do you think we would want to add a hook to sign by `keyctl`, so folks that don't use dracut can benefit as well?

I need to admit that I couldn't take time to read gigantic options of `keyctl`.

> RIght now I'm using https://github.com/Foxboron/sbctl , which I want to package as soon as we get a release.

IIRC, `sbctl` also uses `sbsigntool` to sign EFI binary.
Maybe, things was changed from last time I checked it out.

> Didn't realize we had a hook too, thanks!

I wrote that hook (a bit ugly, though) because I need it ;)
It's able to sign only PE binary (aka kernel), not initrd, which I don't like.
The UEFI bundle is better in this regard.

---
Proof read the comment make me think:
- the sbsign hook should be >50, in order to support generated UEFI bundle.
- the keyctl hook, if any, should have the same index with sbsign, too.

Reply:
> Do you mean dracut's modsign, which uses keyctl?

Don't think so, it's the option that uses `uefi_secureboot_{cert,key}` in [dracut.conf](https://man.voidlinux.org/dracut.conf). It uses `objcopy` to create the UEFI image and then `sbsign` to sign it. I don't know how `keyctl` works.

I opened a PR to launch dracut with `--uefi`, yes. https://github.com/void-linux/void-packages/pull/22484

> If it's not merged, I could take a look next week.

Thanks!

> Creating the bundle doesn't work well with other hooks that is in place, e.g. rEFInd, efimgr. No?

Oof, I don't know, For rEFInd I always drop a `refind_linux.conf` config file in the kernel folder. I've never used the `efibootmgr` hook either.

> IIRC, sbctl also uses sbsigntool to sign EFI binary.

For now, yes. The main advantage is that it takes care of itself, so it generates keys and remembers what it needs to sign, no need to keep X different config files for each application that needs to be signed. For example, in my case I have UEFI bundles, kernel images and rEFInd, and I need to run a single command to sign all of them.

> The UEFI bundle is better in this regard.

Yeah, that's why I started using it c: